### PR TITLE
Add the memory timeline profiling support through the PyTorch profiler.

### PR DESCRIPTION
**Pull request by cli99**
This PR adds the memory timeline profiling support through the PyTorch profiler. 
To enable it, the following flags in the profiler config must be all to set to `true`.
```
  torch_prof_with_stack: true
  torch_prof_record_shapes: true
  torch_prof_profile_memory: true
```

1) By default, the general timeline profiling is also enabled and produces a json trace. To produce memory timeline only, 
explicitly set `torch_prof_filename:` to empty.

2) The default plotting of the memory timeline from PyTorch is not very informative, to use our custom plotting function (with adjusted axis tickers and grids), set `torch_prof_memory_custom_plot: true`

Below is an example memory timeline trace of MPT-7B with micro batch size 4, FSDP full shard,  on 8XH100 (mcli run: 1node-memory-timeline-example-rc5SaM).
Screenshot 2023-12-11 at 7 41 18 PM

### CODE REVIEW

File: composer/profiler/torch_profiler.py [hunk 1]
```diff
@@ -146,30 +184,42 @@ def __init__(
...
     ) -> None:
         self.overwrite = overwrite
         self.folder = folder
-        if use_gzip and not filename.endswith('.gz'):
-            filename += '.gz'
+        if use_gzip:
+            if not filename.endswith('.gz'):
+                filename += '.gz'
+            if memory_filename is not None and not memory_filename.endswith('.html'):
+                memory_filename += '.gz'
```
Comment (j316chuck) [inline composer/profiler/torch_profiler.py:205]: Let's not support gzip html?

File: composer/profiler/torch_profiler.py [hunk 2]
```diff
@@ -146,30 +184,42 @@ def __init__(
...
         self.filename = filename
-        if use_gzip and remote_file_name is not None and not remote_file_name.endswith('.gz'):
-            remote_file_name += '.gz'
+        self.memory_filename = memory_filename
+        if use_gzip:
+            if remote_file_name is not None and not remote_file_name.endswith('.gz'):
+                remote_file_name += '.gz'
+            if memory_remote_file_name is not None and not memory_remote_file_name.endswith('.gz'):
+                memory_remote_file_name += '.gz'
```
Comment (j316chuck) [inline composer/profiler/torch_profiler.py:212]: Can we not support gzip html?

File: composer/profiler/torch_profiler.py [hunk 3]
```diff
@@ -202,28 +252,116 @@ def handler_fn(prof: torch.profiler.profiler.profile):
...
-            state.profiler.record_chrome_json_trace_file(trace_file_name)
-            if self.remote_file_name is not None:
-                trace_remote_file_name = format_name_with_dist_and_time(self.remote_file_name,
-                                                                        run_name=state.run_name,
-                                                                        timestamp=timestamp)
-                trace_remote_file_name = trace_remote_file_name.lstrip('/')
-                logger.upload_file(remote_file_name=trace_remote_file_name,
-                                   file_path=trace_file_name,
-                                   overwrite=self.overwrite)
+            if not (self.filename is None and self.memory_filename is not None):
```
[the remark this answers is not in the document]
Comment (cli99) [reply composer/profiler/torch_profiler.py:255]: by default, it supports both at once. This only evaluates to true when users explictly set `torch_prof_filename` to empty (by default, its not empty if not specified in the yaml).

Comment (j316chuck) [inline composer/profiler/torch_profiler.py:255]:
```suggestion
            if self.filename is not None:
```

We should output traces if self.filename is not None (independent of self.memory_filename)

File: composer/profiler/torch_profiler.py [hunk 4]
```diff
@@ -146,30 +184,42 @@ def __init__(
...
+        memory_filename: Optional[str] = 'rank{rank}.{batch}.pt.trace.memory.html',
+        memory_remote_file_name: Optional[
+            str] = '{run_name}/torch_memory_traces/rank{rank}.{batch}.pt.trace.memory.html',
         overwrite: bool = False,
         use_gzip: bool = False,
         record_shapes: bool = False,
         profile_memory: bool = True,
         with_stack: bool = False,
         with_flops: bool = True,
         num_traces_to_keep: int = -1,
```
[the remark this answers is not in the document]
Comment (cli99) [reply composer/profiler/torch_profiler.py:196]: the current code requires `memory_filename` to be set if users want to enable mem profile. it saves locally with

File: composer/profiler/torch_profiler.py [hunk 5]
```diff
@@ -202,28 +252,116 @@ def handler_fn(prof: torch.profiler.profiler.profile):
...
+                if self.remote_file_name is not None:
+                    trace_remote_file_name = format_name_with_dist_and_time(self.remote_file_name,
+                                                                            run_name=state.run_name,
+                                                                            timestamp=timestamp)
+                    trace_remote_file_name = trace_remote_file_name.lstrip('/')
+                    logger.upload_file(remote_file_name=trace_remote_file_name,
+                                       file_path=trace_file_name,
+                                       overwrite=self.overwrite)
+
+            log.debug(f'Memory profiler enabled: {self.memory_filename}')
```
Comment (j316chuck) [inline composer/profiler/torch_profiler.py:274]:
```suggestion
            log.debug(f'Memory profiler enabled: {self.memory_filename if self.memory_filename else False}')
```

File: composer/profiler/torch_profiler.py [hunk 6]
```diff
@@ -202,28 +252,116 @@ def handler_fn(prof: torch.profiler.profiler.profile):
...
+
+                    import matplotlib.pyplot as plt
+                    import numpy as np
+                    from torch.profiler._memory_profiler import (_CATEGORY_TO_COLORS, _CATEGORY_TO_INDEX,
+                                                                 MemoryProfileTimeline)
+
+                    # Construct the memory timeline plot data
+                    mem_tl = MemoryProfileTimeline(prof._memory_profile())
+
+                    def export_memory_timeline_html(mem_tl, path, device, figsize=(20, 12), title=None) -> None:
```
Comment (j316chuck) [inline composer/profiler/torch_profiler.py:297]: Can we make this it's own util function outside this class

File: composer/profiler/profiler.py [hunk 7]
```diff
@@ -76,6 +76,10 @@ def new_profiler_init(self, dummy_ellipsis=None, **kwargs):
         torch_prof_filename (str, optional): See :class:`~composer.profiler.torch_profiler.TorchProfiler`.
         torch_prof_remote_file_name (str, optional): See :class:`~composer.profiler.torch_profiler.TorchProfiler`.
             Additionally supports full object store paths e.g: s3://bucket/path/to/file.
+        torch_prof_memory_filename (str, optional): See :class:`~composer.profiler.torch_profiler.TorchProfiler`.
+        torch_prof_memory_remote_file_name (str, optional): See :class:`~composer.profiler.torch_profiler.TorchProfiler`.
+            Additionally supports full object store paths e.g: s3://bucket/path/to/file.
+        torch_prof_memory_custom_plot (bool, optional): Whether to plot the memory timeline with grid. (default: ``False``).
```
[the remark this answers is not in the document]
Comment (cli99) [reply composer/profiler/profiler.py:82]: you mean setting `torch_prof_memory_filename (str, optional)` to be not optional and have a default value?
Then by default, when   `torch_prof_with_stack: true
  torch_prof_record_shapes: true
  torch_prof_profile_memory: true`, it always generate a memory timeline trace unless users set `torch_prof_memory_filename` to empty. does this sound okay?

[the remark this answers is not in the document]
Comment (j316chuck) [reply composer/profiler/profiler.py:82]: Apologies, I meant just enable `torch_prof_memory_custom_plot` to be true since your plots look better.

**Comment by j316chuck** [comment]
Thanks for this Cheng! Can we add a mcli run with a memory timeline trace in the PR description for an mpt-7b fsdp model? 

Also can we add two tests, one for ensuring the profiler uploads memory html init /file upload correctly and another to validate the custom plot function you wrote 
 https://github.com/mosaicml/composer/blob/dev/tests/profiler/test_profiler.py#L46-L47.

File: composer/profiler/torch_profiler.py [hunk 8]
```diff
@@ -202,28 +256,60 @@ def handler_fn(prof: torch.profiler.profiler.profile):
...
+                                                                            run_name=state.run_name,
+                                                                            timestamp=timestamp)
+                    trace_remote_file_name = trace_remote_file_name.lstrip('/')
+                    logger.upload_file(remote_file_name=trace_remote_file_name,
+                                       file_path=trace_file_name,
+                                       overwrite=self.overwrite)
+
+            log.debug(f'Memory profiler enabled: {self.memory_filename if self.memory_filename else False}')
+            if self.memory_filename is not None:
+                if version.parse(torch.__version__) > version.parse('2.1.0.dev'):  # type: ignore
```
[the remark this answers is not in the document]
Comment (mvpatel2000) [reply composer/profiler/torch_profiler.py:280]: Let's just require 2.1 given it was released

File: composer/profiler/torch_profiler.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mvpatel2000) [reply composer/profiler/torch_profiler.py:197]: Will this not error as `with_stack` is defualt False?

File: composer/profiler/profiler.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (cli99) [reply composer/profiler/profiler.py]: removed torch_prof_memory_custom_plot

File: composer/profiler/torch_profiler.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (cli99) [reply composer/profiler/torch_profiler.py:286]: @j316chuck  to allow 2.10-rcxx

File: composer/profiler/torch_profiler.py [hunk 9]
```diff
@@ -202,28 +256,60 @@ def handler_fn(prof: torch.profiler.profiler.profile):
...
+                                                       timestamp=timestamp),
+                    )
+                    log.debug(f'Saving memory trace to {memory_trace_file_name}')
+                    memory_trace_file_dirname = os.path.dirname(memory_trace_file_name)
+                    if memory_trace_file_dirname:
+                        os.makedirs(memory_trace_file_dirname, exist_ok=True)
+                    if self.memory_custom_plot:
+                        from composer.profiler.utils import export_memory_timeline_html
+                        export_memory_timeline_html(prof, memory_trace_file_name,
+                                                    torch.cuda.current_device())  # type: ignore
```
[the remark this answers is not in the document]
Comment (cli99) [reply composer/profiler/torch_profiler.py:301]: export_memory_timeline takes `device` as optional[str] but due to  https://github.com/pytorch/pytorch/blob/0692240b90fa7d3c4c43728cbf6a046a64de7725/torch/profiler/profiler.py#L278-L283, we cannot do `str( torch.cuda.current_device()` when we pass in the device value, otherwise the device is not set correctly.

File: composer/profiler/torch_profiler.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (j316chuck) [reply composer/profiler/torch_profiler.py:286]: Got it, I'm okay with keeping 2.1.dev 👍

File: tests/profiler/test_memory_timeline.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    with torch.profiler.profile(record_shapes=True, with_stack=True, profile_memory=True) as prof:
+        y = model(x)
+        loss = torch.nn.functional.mse_loss(y, targets)
+        loss.backward()
+        optimizer.step()
+        optimizer.zero_grad()
+
+    memory_profile = prof._memory_profile()
+    timeline = memory_profile.timeline
+    assert all((t == -1) if action == _memory_profiler.Action.PREEXISTING else (t > 0) for t, action, _, _ in timeline)
```
[the remark this answers is not in the document]
Comment (cli99) [reply tests/profiler/test_memory_timeline.py:40]: this checks the default memory timeline event value (t == -1), https://github.com/pytorch/pytorch/blob/main/torch/profiler/_memory_profiler.py#L715

File: composer/profiler/utils.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        return
+
+    from torch.profiler._memory_profiler import _CATEGORY_TO_COLORS, _CATEGORY_TO_INDEX, MemoryProfileTimeline
+
+    # Default to device 0, if unset. Fallback on cpu.
+    if device is None and prof.use_device and prof.use_device != 'cuda':
+        device = prof.use_device + ':0'
+
+    if device is None:
+        device = 'cuda:0' if torch.cuda.is_available() else 'cpu'
```
[the remark this answers is not in the document]
Comment (cli99) [reply composer/profiler/utils.py:41]: The pytorch default plotting function has the same signature and the `None` check. I think it's useful to keep our plotting function align with the original one

File: tests/profiler/test_memory_timeline.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (j316chuck) [reply tests/profiler/test_memory_timeline.py:40]: Thanks! Can we add that as a comment for readability?

File: composer/profiler/utils.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (j316chuck) [reply composer/profiler/utils.py:41]: Got it, makes sense! Fine to keep for now then

**Review by mvpatel2000 — APPROVED** [review]
Can we move test into `test_profiler`?
